java-client/src/test/java/org/opensearch/client/opensearch/json/JsonpMapperTest.java

```diff
@@ -116,6 +119,60 @@ public void testJacksonCustomJsonFactory() {
         IOUtils.closeWhileHandlingException(writer);
     }
 
+    @Test
+    public void testConcurrentLazyResolve() throws Exception {
+        // Test fix for issue 172 - concurrency error in LazyDeserializer
+        // This latch holds off resolution of the LazyDeserializer until we're sure that
+        // two threads are attempting it simultaneously
+        CountDownLatch trigger = new CountDownLatch(1);
+        JsonpDeserializer<Integer> deserializer = JsonpDeserializer.lazy(() -> {
+            try {
+                trigger.await();
+            } catch (Exception e) {
+                throw new RuntimeException("Interrupted", e);
+            }
+            return JsonpDeserializer.integerDeserializer();
+        });
+
+        // Two threads will attempt to deserialize.  They should both be successful
+        final AtomicInteger successes = new AtomicInteger(0);
+        final JsonpMapper mapper = new JsonbJsonpMapper();
+        Runnable threadProc = () -> {
+            JsonParser parser = mapper.jsonProvider().createParser(new StringReader("0"));
+            try {
+                // Prior to fix, one of these would throw NPE because its
+                // LazyDeserializer resolution would return null
+                deserializer.deserialize(parser,mapper);
+                successes.incrementAndGet();
+            } catch (Throwable e) {
+                // We'll notice that we failed to increment successes
+            }
+        };
+        // Two identical threads
+        Thread thread1 = new Thread(threadProc);
+        thread1.setDaemon(true);
+        thread1.start();
+        Thread thread2 = new Thread(threadProc);
+        thread2.setDaemon(true);
+        thread2.start();
+
+        // Wait until both threads are blocked waiting LazyDeserializer resolution
+        do {
+            try {
+                Thread.sleep(5);
```

Comment:
Does it really take 5 seconds? Doubt it, maybe this should be very short to avoid hanging the test unnecessarily?

Reply:
Oh 🤦 long day :)